### PR TITLE
Add docs for asset caching x-script usage

### PR DESCRIPTION
**Pull request by dan-shaw**

### CODE REVIEW

File: vcpkg/users/assetcaching.md [hunk 1]
```diff
@@ -57,3 +57,70 @@ See also the [binary caching documentation for Azure Blob Storage](binarycaching
 Syntax: `x-block-origin`
 
 Disables use of the original URLs in case the mirror does not have the file available.
+
+#### `x-script`
+
```
Comment (ras0219-msft) [inline vcpkg/users/assetcaching.md:62]: This should have a `Syntax:` line like the other two.

File: vcpkg/users/assetcaching.md [hunk 2]
```diff
@@ -57,3 +57,70 @@ See also the [binary caching documentation for Azure Blob Storage](binarycaching
...
+Dispatches to an external tool to fetch the asset. Within the template, `{url}` will be replaced by the original url, `{sha512}` will be replaced by the SHA512 value, and `{dst}` will be replaced by the output path to save to. These substitutions will all be properly shell escaped, so an example template would be: `curl -L {url} --output {dst}`. `{{` will be replaced by `}` and `}}` will be replaced by `}` to avoid expansion. Note that this will be executed inside the build environment, so the PATH and other environment variables will be modified by the triplet.
+
+## Example
```
Comment (ras0219-msft) [inline vcpkg/users/assetcaching.md:65]: Markdown headers (lines beginning with `#`) are hierarchical -- this means that this example section needs to be
```suggestion
##### Example
```
So it's "under" the `x-script` section.

Comment (ras0219-msft) [inline vcpkg/users/assetcaching.md:65]: Considering the length of this example, I think we should split it out into a separate page underneath `Examples` and link to it from here.

File: vcpkg/examples/assetcaching-script.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+In this example, we will set up a NuGet feed as a source, meaning we will use `nuget.exe` to pull and push assets.
+
+First, we will set up the environment before any call to `vcpkg install`:
+```powershell
+$env:X_VCPKG_ASSET_SOURCES="clear;x-script,C:/path/to/sources.bat {url} {sha512} {dst};x-block-origin"
+$env:NUGET=C:/path/to/nuget.exe
+$env:VCPKG_KEEP_ENV_VARS=NUGET
+```
+`X_VCPKG_ASSET_SOURCES` passes in a script `sources.bat` which will handle `nuget` push and pull commands. The `vcpkg` environment needs to use a copy of `nuget.exe` which can be supplied by the environment with [VCPKG_KEEP_ENV_VARS](config-environment.md#vcpkg_keep_env_vars).
```
Comment (ras0219-msft) [inline vcpkg/examples/assetcaching-script.md:21]:
```suggestion
[`X_VCPKG_ASSET_SOURCES`](../users/config-environment.md#x_vcpkg_asset_sources) defines a single [`x-script`](../users/assetcaching.md#x-script) asset source while blocking fallback to the original url ([`x-block-origin`](../users/assetcaching.md#x-block-origin)). The `x-script` source is defined to call a script `sources.bat` which will handle `nuget` push and pull commands. This script must be able to run both inside and outside the inner build environment, so we pass it the path to `nuget.exe` in the environment with `NUGET` and [`VCPKG_KEEP_ENV_VARS`](config-environment.md#vcpkg_keep_env_vars).
```

File: vcpkg/examples/assetcaching-script.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+[!INCLUDE [experimental](../../includes/experimental.md)]
+
+Your scenario might need a different storage source such as NuGet feed, Universal Packages, or something else.
+
+In this example, we will set up a NuGet feed as a source, meaning we will use `nuget.exe` to pull and push assets.
+
+First, we will set up the environment before any call to `vcpkg install`:
+```powershell
+$env:X_VCPKG_ASSET_SOURCES="clear;x-script,C:/path/to/sources.bat {url} {sha512} {dst};x-block-origin"
```
Comment (ras0219-msft) [inline vcpkg/examples/assetcaching-script.md:17]: I think `sources.bat` is a bit of a confusing name; maybe `assetprovider.bat`?

File: vcpkg/examples/assetcaching-script.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+$env:NUGET=C:/path/to/nuget.exe
+$env:VCPKG_KEEP_ENV_VARS=NUGET
+```
+`X_VCPKG_ASSET_SOURCES` passes in a script `sources.bat` which will handle `nuget` push and pull commands. The `vcpkg` environment needs to use a copy of `nuget.exe` which can be supplied by the environment with [VCPKG_KEEP_ENV_VARS](config-environment.md#vcpkg_keep_env_vars).
+
+The script `sources.bat` checks if the asset is stored in the feed, and if so, installs it to the destination path. If the asset is not stored on the feed, the script fetches the url via curl, packs the asset in a `nuget` package and pushes it to the feed.
+
+
+`sources.bat`
```
Comment (ras0219-msft) [inline vcpkg/examples/assetcaching-script.md:26]: As a point of style, we should not introduce files by simply having their name in backticks and then the content -- instead, we should use prose to instruct the user to create a file name so-and-so.

(example from the dotnet docs: https://learn.microsoft.com/en-us/aspnet/core/tutorials/min-web-api?view=aspnetcore-7.0&tabs=visual-studio#the-model-and-database-context-classes)

```suggestion
Create a `sources.bat` file in your project with the following content:
```

This should probably be reordered above the environment setting steps above

File: vcpkg/examples/assetcaching-script.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
+---
+title: Asset Caching with an alternate storage source
```
Comment (ras0219-msft) [inline vcpkg/examples/assetcaching-script.md:2]: I think this example should focus more on being a "How To" or a "Tutorial". The objective of a How To article is to first promise the user where they will end up (with an `x-script` asset cache pointing at NuGet in ADO) and then successfully transport them there (via concrete steps: "First, do X." "Then, do Y.").

(consider browsing https://diataxis.fr/how-to-guides/ for more philosophical inspiration)

```suggestion
title: How to create an x-script Asset Caching source for NuGet
```

File: vcpkg/examples/assetcaching-script.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+description: Learn about adding a new storage source
+ms.date: 4/18/2023
+---
+
+# Asset Caching with an alternate storage source
+
+[!INCLUDE [experimental](../../includes/experimental.md)]
+
+Your scenario might need a different storage source such as NuGet feed, Universal Packages, or something else.
```
Comment (ras0219-msft) [inline vcpkg/examples/assetcaching-script.md:11]: As an continuation of the above point, a How-To shouldn't distract the user with other possibilities.

File: vcpkg/users/assetcaching.md [hunk 8]
```diff
@@ -57,3 +57,13 @@ See also the [binary caching documentation for Azure Blob Storage](binarycaching
...
+
+#### `x-script`
+
+Syntax: `x-script,<template>`
+
+Dispatches to an external tool to fetch the asset. Within the template, `{url}` will be replaced by the original url, `{sha512}` will be replaced by the SHA512 value, and `{dst}` will be replaced by the output path to save to. These substitutions will all be properly shell escaped, so an example template would be: `curl -L {url} --output {dst}`. `{{` will be replaced by `}` and `}}` will be replaced by `}` to avoid expansion. Note that this will be executed inside the build environment, so the PATH and other environment variables will be modified by the triplet.
+
+##### Examples
+
+This [example](../examples/assetcaching-script.md) uses `x-script` to support a different source than Azure Blob Storage.
```
Comment (ras0219-msft) [inline vcpkg/users/assetcaching.md:69]:
```suggestion
See [How to create an x-script Asset Caching source for NuGet](../examples/assetcaching-script.md) for a walkthrough using `x-script`.
```

File: vcpkg/examples/assetcaching-script.md (hunk 6 above)
Comment (ras0219-msft) [reply vcpkg/examples/assetcaching-script.md:2]: If you change the title to reflect it being a how-to, you probably should also change the file name.

File: vcpkg/examples/assetcaching-script.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    xcopy /f C:\path\to\test.nuspec .
+    echo %_dst%
+    xcopy /f %_dst% .
+
+    %NUGET% pack test.nuspec -Properties sha=%_sha512%
+    %NUGET% push %_sha512%.1.0.0.nupkg -Source <feed url>
+)
+```
+
+`test.nuspec`
```
Comment (ras0219-msft) [inline vcpkg/examples/assetcaching-script.md:57]:
```suggestion
`asset-source.nuspec`
```

File: vcpkg/examples/assetcaching-script.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    curl.exe -L %url% --create-dirs --output %dst%
+    cd %dst%/../
+    mkdir tempdir
+    cd tempdir
+    xcopy /f C:\path\to\test.nuspec .
+    echo %_dst%
+    xcopy /f %_dst% .
+
+    %NUGET% pack test.nuspec -Properties sha=%_sha512%
+    %NUGET% push %_sha512%.1.0.0.nupkg -Source <feed url>
```
Comment (ras0219-msft) [inline vcpkg/examples/assetcaching-script.md:53]:
```suggestion
    %NUGET% pack C:\path\to\test.nuspec -BasePath %~dp3 -Properties "sha=%_sha512%;file=%dst%" -OutputDirectory %TEMP%
    %NUGET% push %TEMP%\%_sha512%.1.0.0.nupkg -Source <feed url>
```

File: vcpkg/examples/assetcaching-script.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+`sources.bat`
+```bat
+set url=%1
+set sha512=%2
+set dst=%3
+set "_dst=%dst:/=\%"
+set "_sha512=%sha512:~0,90%"
+
+cd %dst%/../
```
Comment (ras0219-msft) [inline vcpkg/examples/assetcaching-script.md:34]:
```suggestion
cd /d %~dp3
```
Doesn't need to resolve `/../` and successfully changes drives!

File: vcpkg/examples/assetcaching-script.md [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+if exist %_sha512%.1.0.0 (
+    echo "Pull from the NuGet feed"
+    cd %_sha512%.1.0.0
+
+    REM Assume both are files not directories
+    echo "F" | xcopy /f *.part %_dst%
+) else (
+    echo "Fetch from the url"
+    curl.exe -L %url% --create-dirs --output %dst%
+    cd %dst%/../
```
Comment (ras0219-msft) [inline vcpkg/examples/assetcaching-script.md:45]: You already `cd`'d above.

**Comment by dan-shaw** [comment]
also, I've tested the new script locally and it works!

**Review by ras0219-msft — APPROVED** [review]

**Pull request merged**